File: code/programs/python/unix-tools/seq_tool.py
"""seq — print a sequence of numbers.

=== What This Program Does ===

This is a reimplementation of the GNU ``seq`` utility. It generates a
sequence of numbers from FIRST to LAST with an optional INCREMENT, printing
each number on its own line (or separated by a custom string).

=== Invocation Forms ===

seq supports three argument patterns:

1. ``seq LAST``           — count from 1 to LAST by 1
2. ``seq FIRST LAST``     — count from FIRST to LAST by 1
3. ``seq FIRST INCR LAST`` — count from FIRST to LAST by INCR

For example::

    $ seq 5          # prints 1 2 3 4 5
    $ seq 2 5        # prints 2 3 4 5
    $ seq 1 2 10     # prints 1 3 5 7 9

=== Floating Point Support ===

seq handles both integers and floating-point numbers. If any argument
contains a decimal point, the output uses floating-point formatting::

    $ seq 0.5 0.5 2.5
    0.5
    1.0
    1.5
    2.0
    2.5

=== Equal Width (-w) ===

The ``-w`` flag pads numbers with leading zeros so they all have the
same width. This is useful for generating filenames::

    $ seq -w 1 100
    001
    002
    ...
    100

=== Custom Format (-f) ===

The ``-f`` flag accepts a printf-style format string::

    $ seq -f "file_%03g" 1 3
    file_001
    file_002
    file_003

=== Custom Separator (-s) ===

The ``-s`` flag changes the separator between numbers (default is newline)::

    $ seq -s ", " 1 5
    1, 2, 3, 4, 5

=== CLI Builder Integration ===

The entire CLI is defined in ``seq.json``. CLI Builder handles flag parsing,
help text, and version output. This file implements the sequence generation
and formatting logic.
"""

from __future__ import annotations

import math
import sys
from decimal import Decimal, InvalidOperation
from pathlib import Path

# ---------------------------------------------------------------------------
# Locate the JSON spec file.
# ---------------------------------------------------------------------------

SPEC_FILE = str(Path(__file__).parent / "seq.json")


def parse_number(value: str) -> Decimal:
    """Parse a string into a Decimal number.

    We use Decimal rather than float to avoid floating-point precision
    issues. For example, with floats, ``0.1 + 0.1 + 0.1`` might not
    equal ``0.3`` exactly. Decimal arithmetic preserves the precision
    that the user specified.

    Args:
        value: The string representation of the number.

    Returns:
        A Decimal value.

    Raises:
        SystemExit: If the value is not a valid number.
    """
    try:
        return Decimal(value)
    except InvalidOperation:
        print(f"seq: invalid floating point argument: '{value}'", file=sys.stderr)
        raise SystemExit(1) from None


def decimal_places(value: Decimal) -> int:
    """Count the number of decimal places in a Decimal value.

    This is used to determine output precision. If the user writes
    ``seq 0.5 0.5 2.0``, we want output like ``0.5``, ``1.0``, etc.

    The number of decimal places is determined by the exponent:
    - ``Decimal('1')`` has exponent 0 -> 0 decimal places
    - ``Decimal('1.5')`` has exponent -1 -> 1 decimal place
    - ``Decimal('1.50')`` has exponent -2 -> 2 decimal places

    Args:
        value: The Decimal to examine.

    Returns:
        The number of decimal places (0 if integer).
    """
    # The sign, digits, and exponent of the Decimal.
    _, _, exp = value.as_tuple()
    if isinstance(exp, str):
        # Special values like Infinity or NaN.
        return 0
    return max(0, -exp)


def format_number(
    value: Decimal,
    *,
    precision: int,
    width: int,
    equal_width: bool,
    fmt: str | None,
) -> str:
    """Format a number for output.

    This function handles three formatting modes:

    1. **Custom format** (``-f``): Use a printf-style format string.
       The format uses ``%g``, ``%f``, or ``%e`` specifiers.

    2. **Equal width** (``-w``): Pad with leading zeros to the specified
       width. The width is determined by the widest number in the sequence.

    3. **Default**: Use the natural precision of the input numbers.

    Args:
        value: The number to format.
        precision: Number of decimal places for default formatting.
        width: Minimum width for equal-width mode.
        equal_width: Whether to pad with leading zeros.
        fmt: Optional printf-style format string.

    Returns:
        The formatted number string.
    """
    if fmt is not None:
        # Convert printf-style format (%g, %f, %e) to Python format.
        # We replace %g/%f/%e with Python's equivalent.
        return fmt % float(value)

    if precision == 0:
        num_str = str(int(value))
    else:
        num_str = f"{float(value):.{precision}f}"

    if equal_width:
        # Pad with leading zeros. Handle negative numbers correctly:
        # the minus sign should come before the zeros.
        if value < 0:
            return "-" + num_str[1:].zfill(width - 1)
        return num_str.zfill(width)

    return num_str


def generate_sequence(
    first: Decimal,
    increment: Decimal,
    last: Decimal,
) -> list[Decimal]:
    """Generate a sequence of numbers from first to last by increment.

    The sequence includes ``first`` and includes ``last`` if the sequence
    reaches it exactly. The direction is determined by the sign of
    increment:

    - Positive increment: sequence goes up (first must be <= last)
    - Negative increment: sequence goes down (first must be >= last)
    - Zero increment: infinite loop (we cap at a reasonable limit)

    Args:
        first: The starting number.
        increment: The step between numbers.
        last: The ending number (inclusive if reached exactly).

    Returns:
        A list of Decimal values in the sequence.
    """
    result: list[Decimal] = []

    if increment == 0:
        # GNU seq with increment 0 runs forever. We don't do that.
        return result

    if increment > 0:
        current = first
        while current <= last:
            result.append(current)
            current += increment
    else:
        current = first
        while current >= last:
            result.append(current)
            current += increment

    return result


def main() -> None:
    """Entry point: parse args via CLI Builder, then generate sequence."""
    from cli_builder import HelpResult, ParseErrors, Parser, ParseResult, VersionResult

    # --- Step 1: Parse arguments -------------------------------------------
    try:
        result = Parser(SPEC_FILE, sys.argv).parse()
    except ParseErrors as exc:
        for error in exc.errors:
            print(f"seq: {error.message}", file=sys.stderr)
        raise SystemExit(1) from None

    # --- Step 2: Dispatch on result type -----------------------------------
    if isinstance(result, HelpResult):
        print(result.text)
        raise SystemExit(0)

    if isinstance(result, VersionResult):
        print(result.version)
        raise SystemExit(0)

    # --- Step 3: Business logic --------------------------------------------
    assert isinstance(result, ParseResult)

    separator = result.flags.get("separator", "\n")
    equal_width = result.flags.get("equal_width", False)
    fmt = result.flags.get("format")

    # Parse the positional arguments: LAST, or FIRST LAST, or FIRST INCR LAST.
    numbers = result.arguments.get("numbers", [])
    if isinstance(numbers, str):
        numbers = [numbers]

    if len(numbers) == 1:
        first = Decimal("1")
        increment = Decimal("1")
        last = parse_number(numbers[0])
    elif len(numbers) == 2:  # noqa: PLR2004
        first = parse_number(numbers[0])
        increment = Decimal("1")
        last = parse_number(numbers[1])
    elif len(numbers) == 3:  # noqa: PLR2004
        first = parse_number(numbers[0])
        increment = parse_number(numbers[1])
        last = parse_number(numbers[2])
    else:
        print("seq: missing operand", file=sys.stderr)
        raise SystemExit(1)

    # Determine output precision from the input values.
    precision = max(
        decimal_places(first),
        decimal_places(increment),
        decimal_places(last),
    )

    # Generate the sequence.
    sequence = generate_sequence(first, increment, last)

    if not sequence:
        return

    # Determine width for equal-width mode.
    width = 0
    if equal_width:
        width = max(len(format_number(n, precision=precision, width=0, equal_width=False, fmt=None)) for n in sequence)

    # Format and output.
    formatted = [
        format_number(n, precision=precision, width=width, equal_width=equal_width, fmt=fmt)
        for n in sequence
    ]

    print(separator.join(formatted))


if __name__ == "__main__":
    main()

File: code/programs/ruby/unix-tools/seq_tool.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# seq_tool.rb -- Print a sequence of numbers
# ============================================
#
# === What This Program Does ===
#
# This is a reimplementation of the GNU `seq` utility. It prints a
# sequence of numbers from FIRST to LAST, separated by a string
# (default: newline).
#
# === Argument Forms ===
#
# seq accepts 1, 2, or 3 positional arguments:
#
#   seq LAST              -> prints 1 to LAST (increment 1)
#   seq FIRST LAST        -> prints FIRST to LAST (increment 1 or -1)
#   seq FIRST INCR LAST   -> prints FIRST to LAST (increment INCR)
#
# The numbers can be integers or floating-point:
#
#     $ seq 5
#     1 2 3 4 5
#
#     $ seq 2 0.5 4
#     2.0 2.5 3.0 3.5 4.0
#
# === Equal Width (-w) ===
#
# The -w flag pads all numbers with leading zeros so they are the
# same width:
#
#     $ seq -w 8 10
#     08
#     09
#     10
#
# === Custom Separator (-s) ===
#
# By default, numbers are separated by newlines. The -s flag lets
# you choose a different separator:
#
#     $ seq -s ", " 3
#     1, 2, 3
#
# The output always ends with a newline, regardless of the separator.
#
# === Format (-f) ===
#
# The -f flag lets you specify a printf-style format string:
#
#     $ seq -f "Item %03g" 3
#     Item 001
#     Item 002
#     Item 003

require "coding_adventures_cli_builder"

# ---------------------------------------------------------------------------
# Locate the JSON spec file.
# ---------------------------------------------------------------------------

SEQ_SPEC_FILE = File.join(File.dirname(__FILE__), "seq.json")

# ---------------------------------------------------------------------------
# Business Logic: parse_seq_args
# ---------------------------------------------------------------------------
# Parse the 1-3 positional arguments into first, increment, last.
#
# seq LAST           -> first=1, incr=1, last=LAST
# seq FIRST LAST     -> first=FIRST, incr=1 (or -1 if FIRST > LAST), last=LAST
# seq FIRST INCR LAST -> first=FIRST, incr=INCR, last=LAST
#
# Returns: [Float, Float, Float] -- first, increment, last

def parse_seq_args(numbers)
  case numbers.length
  when 1
    [1.0, 1.0, Float(numbers[0])]
  when 2
    first = Float(numbers[0])
    last = Float(numbers[1])
    incr = first <= last ? 1.0 : -1.0
    [first, incr, last]
  when 3
    [Float(numbers[0]), Float(numbers[1]), Float(numbers[2])]
  else
    raise ArgumentError, "seq: expected 1-3 arguments, got #{numbers.length}"
  end
end

# ---------------------------------------------------------------------------
# Business Logic: format_number
# ---------------------------------------------------------------------------
# Format a number for output. If the number is an integer value (no
# fractional part), format it as an integer. Otherwise, format as float.
# This matches GNU seq behavior: `seq 3` prints "1\n2\n3" not "1.0\n2.0\n3.0".
#
# Parameters:
#   num        - The number to format
#   fmt        - Optional printf format string (overrides auto-detection)
#   pad_width  - Width for zero-padding (0 for no padding)
#
# Returns: The formatted number string.

def format_seq_number(num, fmt, pad_width)
  if fmt
    # User-specified format string.
    format(fmt, num)
  elsif num == num.floor && num.abs < 1e15
    # Integer value: format without decimal point.
    str = num.to_i.to_s
    if pad_width > 0
      # Zero-pad to the required width. Negative numbers get special
      # treatment: the minus sign counts toward the width.
      if num < 0
        "-" + str[1..].rjust(pad_width - 1, "0")
      else
        str.rjust(pad_width, "0")
      end
    else
      str
    end
  else
    # Floating-point value: use %g for clean output.
    str = format("%g", num)
    if pad_width > 0
      if num < 0
        "-" + str[1..].rjust(pad_width - 1, "0")
      else
        str.rjust(pad_width, "0")
      end
    else
      str
    end
  end
end

# ---------------------------------------------------------------------------
# Business Logic: compute_pad_width
# ---------------------------------------------------------------------------
# Compute the width needed for zero-padding with -w.
#
# The width is determined by the widest number that will be printed,
# which is either the first or last value.

def compute_pad_width(first, last)
  # Format first and last as they would appear (without padding) and
  # take the longer one's width.
  w1 = if first == first.floor && first.abs < 1e15
         first.to_i.to_s.length
       else
         format("%g", first).length
       end
  w2 = if last == last.floor && last.abs < 1e15
         last.to_i.to_s.length
       else
         format("%g", last).length
       end
  [w1, w2].max
end

# ---------------------------------------------------------------------------
# Main entry point
# ---------------------------------------------------------------------------

def seq_main
  # --- Step 1: Parse arguments ---------------------------------------------
  begin
    result = CodingAdventures::CliBuilder::Parser.new(SEQ_SPEC_FILE, ["seq"] + ARGV).parse
  rescue CodingAdventures::CliBuilder::ParseErrors => e
    e.errors.each { |err| warn "seq: #{err.message}" }
    exit 1
  end

  # --- Step 2: Dispatch on result type -------------------------------------
  case result
  when CodingAdventures::CliBuilder::HelpResult
    puts result.text
    exit 0
  when CodingAdventures::CliBuilder::VersionResult
    puts result.version
    exit 0
  when CodingAdventures::CliBuilder::ParseResult
    # --- Step 3: Business logic --------------------------------------------
    numbers = result.arguments.fetch("numbers", [])
    separator = result.flags["separator"] || "\n"
    equal_width = result.flags["equal_width"]
    fmt = result.flags["format"]

    begin
      first, incr, last = parse_seq_args(numbers)
    rescue ArgumentError => e
      warn e.message
      exit 1
    end

    # Cannot iterate with zero increment.
    if incr == 0.0
      warn "seq: zero increment"
      exit 1
    end

    # Compute zero-padding width if -w is set.
    pad_width = equal_width ? compute_pad_width(first, last) : 0

    # Generate the sequence. We collect into an array first so we can
    # join with the separator, then print with a trailing newline.
    #
    # The loop condition depends on the direction of the increment:
    #   - Positive increment: keep going while current <= last
    #   - Negative increment: keep going while current >= last
    #
    # We add a small epsilon tolerance to handle floating-point
    # rounding errors (e.g., 0.1 + 0.1 + 0.1 might be 0.30000000000000004).
    values = []
    current = first
    epsilon = (incr.abs * 1e-10)

    if incr > 0
      while current <= last + epsilon
        values << format_seq_number(current, fmt, pad_width)
        current += incr
      end
    else
      while current >= last - epsilon
        values << format_seq_number(current, fmt, pad_width)
        current += incr
      end
    end

    print values.join(separator)
    puts
  end
end

# Only run main when this file is executed directly.
seq_main if __FILE__ == $PROGRAM_NAME
